WalletWasabi/Tor/TorMonitor.cs

```diff
@@ -1,38 +1,133 @@
+		private CancellationTokenSource LoopCts { get; } = new();
+
+		public static bool RequestFallbackAddressUsage { get; private set; }
 		private Uri FallbackBackendUri { get; }
 		private TorHttpClient HttpClient { get; }
 		private TorProcessManager TorProcessManager { get; }
 
+		public async Task StartBootstrapMonitorAsync(CancellationToken appShutdownToken)
+		{
+			using var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(appShutdownToken, LoopCts.Token);
+
+			while (!linkedCts.IsCancellationRequested)
+			{
+				try
+				{
+					(CancellationToken cancellationToken, TorControlClient client) = await TorProcessManager.WaitForNextAttemptAsync(linkedCts.Token).ConfigureAwait(false);
+
+					Logger.LogInfo("Starting Tor bootstrap monitor…");
+
+					List<string> eventNames = new()
+					{
+						StatusEvent.EventNameStatusGeneral,
+						StatusEvent.EventNameStatusClient,
+						StatusEvent.EventNameStatusServer,
+						CircEvent.EventName,
+						OrConnEvent.EventName,
+						NetworkLivenessEvent.EventName,
+					};
+					await client.SubscribeEventsAsync(eventNames, cancellationToken).ConfigureAwait(false);
+
+					bool circuitEstablished = false;
+
+					await foreach (TorControlReply reply in client.ReadEventsAsync(cancellationToken).ConfigureAwait(false))
+					{
+						IAsyncEvent asyncEvent;
+
+						try
+						{
+							asyncEvent = AsyncEventParser.Parse(reply);
+						}
+						catch (TorControlReplyParseException e)
+						{
+							Logger.LogError($"Exception thrown when parsing event: '{reply}'", e);
+							continue;
+						}
+
+						if (asyncEvent is BootstrapStatusEvent bootstrapEvent)
+						{
+							BootstrapStatusEvent.Phases.TryGetValue(bootstrapEvent.Progress, out string? bootstrapInfo);
+							Logger.LogInfo($"Bootstrap progress: {bootstrapEvent.Progress}/100 ({bootstrapInfo ?? "N/A"})");
+						}
+						else if (asyncEvent is StatusEvent statusEvent)
+						{
+							if (statusEvent.Action == StatusEvent.ActionCircuitEstablished)
+							{
+								if (!circuitEstablished)
+								{
+									Logger.LogInfo("Tor circuit was established.");
+									circuitEstablished = true;
+								}
+							}
+						}
+						else if (asyncEvent is CircEvent circEvent)
+						{
+							CircuitInfo info = circEvent.CircuitInfo;
+
+							if (info.CircStatus is CircStatus.BUILT or CircStatus.EXTENDED or CircStatus.GUARD_WAIT)
+							{
+								if (!circuitEstablished)
+								{
+									Logger.LogInfo("Tor circuit was established.");
+									circuitEstablished = true;
+								}
+							}
+						}
+					}
+				}
+				catch (OperationCanceledException)
+				{
+					if (linkedCts.IsCancellationRequested)
+					{
+						Logger.LogDebug("Tor Monitor is stopping.");
+					}
+					else
+					{
+						Logger.LogDebug("Tor Monitor is re-initializing.");
+					}
+				}
+				catch (Exception e)
+				{
+					Logger.LogError(e);
+					break;
+				}
+			}
+		}
```

Comment:
This doesn't feel good. Maybe it is, but can't it be inside the periodic runner's boiler plate instead? Or maybe even create a new periodic runner for this, idk, something smells I think.